client/peerset/src/protocol_controller.rs

```diff
@@ -0,0 +1,559 @@
+		match state {
+			PeerState::Connected(Direction::Inbound) => self.num_in -= 1,
+			PeerState::Connected(Direction::Outbound) => self.num_out -= 1,
+			PeerState::NotConnected => self.alloc_slots(),
+		}
+	}
+
+	fn on_remove_reserved_peer(&mut self, peer_id: PeerId) {
+		let mut state = match self.reserved_nodes.remove(&peer_id) {
+			Some(state) => state,
+			None => return,
+		};
+
+		match state {
+			PeerState::Connected(d) => {
+				if self.reserved_only {
+					// Disconnect the node.
+					info!(
+						target: "peerset",
+						"Disconnecting previously reserved node {} on {:?}.",
+						peer_id, self.set_id
+					);
+					state = PeerState::NotConnected;
+					self.drop_connection(peer_id);
+				} else {
+					// Count connections as of regular node.
+					match d {
+						Direction::Inbound => self.num_in += 1,
+						Direction::Outbound => self.num_out += 1,
+					}
+				}
+			},
+			PeerState::NotConnected => {},
+		}
+
+		// Put the node into the list of regular nodes.
+		let prev = self.nodes.insert(peer_id, state);
+		assert!(prev.is_none(), "Corrupted state: reserved node was also non-reserved.");
+	}
+
+	fn on_set_reserved_peers(&mut self, peer_ids: HashSet<PeerId>) {
+		// Determine the difference between the current group and the new list.
+		let current = self.reserved_nodes.keys().cloned().collect();
+		let to_insert = peer_ids.difference(&current).cloned().collect::<Vec<_>>();
+		let to_remove = current.difference(&peer_ids).cloned().collect::<Vec<_>>();
+
+		for node in to_insert {
+			self.on_add_reserved_peer(node);
+		}
+
+		for node in to_remove {
+			self.on_remove_reserved_peer(node);
+		}
+	}
+
+	fn on_set_reserved_only(&mut self, reserved_only: bool) {
+		self.reserved_only = reserved_only;
+
+		if !reserved_only {
+			return self.alloc_slots()
+		}
+
+		// Disconnect all non-reserved peers.
+		self.nodes
+			.iter_mut()
+			.filter_map(|(peer_id, state)| {
+				state.is_connected().then_some({
+					*state = PeerState::NotConnected;
+					*peer_id
+				})
+			})
+			.collect::<Vec<_>>()
+			.iter()
+			.for_each(|peer_id| self.drop_connection(*peer_id));
+	}
+
+	fn on_add_peer(&mut self, peer_id: PeerId) {
+		if self.reserved_nodes.contains_key(&peer_id) {
+			return
+		}
+
+		if self.nodes.insert(peer_id, PeerState::NotConnected).is_none() {
+			self.alloc_slots();
+		}
+	}
+
+	fn on_remove_peer(&mut self, peer_id: PeerId) {
+		// Don't do anything if the node is reserved.
+		if self.reserved_nodes.contains_key(&peer_id) {
+			return
+		}
+
+		match self.nodes.remove(&peer_id) {
+			Some(state) => match state {
+				PeerState::Connected(d) => {
+					match d {
+						Direction::Inbound => self.num_in -= 1,
+						Direction::Outbound => self.num_out -= 1,
+					}
+					self.drop_connection(peer_id);
+				},
+				PeerState::NotConnected => {},
+			},
+			None => {
+				trace!(
+					target: "peerset",
+					"Trying to remove unknown peer {} from {:?}",
+					peer_id, self.set_id,
+				);
+			},
+		}
```

Comment:
```suggestion
        if let Some(PeerState::Connected(d)) = self.nodes.remove(&peer_id) {
            match d {
                Direction::Inbound => self.num_in -= 1,
                Direction::Outbound => self.num_out -= 1,
            }
            self.drop_connection(peer_id);
        } else {
            trace!(
                target: "peerset",
                "Trying to remove unknown peer {} from {:?}",
                peer_id, self.set_id,
            );
        }
```

Reply:
yes good point, I'm not sure if the `trace!()` itself is very valuable but let's keep it as it was